rubix-core/src/main/java/com/qubole/rubix/core/NonLocalReadRequestChain.java

```diff
@@ -12,24 +12,115 @@
     {
-        super(inputStream);
+        this.remoteNodeName = remoteNodeName;
+        this.fs = fs;
+        this.conf = conf;
+        this.strictMode = CacheConfig.isStrictMode(conf);
+        this.filePath = remotePath;
     }
 
     public ReadRequestChainStats getStats()
     {
         return new ReadRequestChainStats()
-                .setRemoteReads(requests)
                 .setNonLocalReads(requests)
                 .setRequestedRead(totalRead)
                 .setNonLocalDataRead(totalRead);
     }
+
+    @Override
+    public Integer call()
+            throws Exception
+    {
+        if (readRequests.size() == 0) {
+            return 0;
+        }
+
+        try {
+            this.bookKeeperClient = createBookKeeperClient(remoteNodeName, conf);
+        }
+        catch (Exception e) {
+            if (strictMode) {
+                throw Throwables.propagate(e);
+            }
+            log.warn("Could not create BookKeeper Client " + Throwables.getStackTraceAsString(e));
+            return directReadRequest(0);
+        }
+
+        for (ReadRequest readRequest : readRequests) {
+            int readLength = 0;
+            int lengthRemaining = readRequest.getActualReadLength();
+            int bufferLength = CacheConfig.getBufferSize(conf);
+
+            DataRead dataRead;
+
+            while (lengthRemaining > 0) {
+                if (lengthRemaining < bufferLength) {
+                    bufferLength = lengthRemaining;
+                }
+                try {
+                    dataRead = bookKeeperClient.readData(filePath, readRequest.getActualReadStart() + readLength, readRequest.getDestBufferOffset(), bufferLength);
+                }
+                catch (TException e) {
+                    return directReadRequest(readRequests.indexOf(readRequest));
+                }
+                System.arraycopy(dataRead.getData(), 0, readRequest.destBuffer, readRequest.getDestBufferOffset() + readLength, dataRead.getSizeRead());
+                readLength += dataRead.getSizeRead();
+                lengthRemaining = readRequest.getActualReadLength() - readLength;
+                totalRead += dataRead.getSizeRead();
+            }
+        }
+
+        if (bookKeeperClient != null) {
+            bookKeeperClient.close();
+        }
+
+        log.info(String.format("Read %d bytes directly from node %s", totalRead, remoteNodeName));
+        return totalRead;
+    }
+
+    private int directReadRequest(int index)
+            throws Exception
+    {
+        int read;
+        inputStream = fs.open(new Path(filePath));
+        DirectReadRequestChain readChain = new DirectReadRequestChain(inputStream);
+        for (ReadRequest readRequest : readRequests.subList(index, readRequests.size())) {
+            readChain.addReadRequest(readRequest);
+        }
+        readChain.lock();
+        read = readChain.call();
+        return read;
```

Comment:
`return read` can be wrong in some cases. We need to return the data size read from DirectRR + the data size read before `index` via NonLocalRR